frontend/test/pytest/test_jax_dynamic_api.py

```diff
@@ -0,0 +1,387 @@
+    dtype = complex
+
+    @qjit
+    def f(s):
+        return jnp.ones(s, dtype)[0, 1, 0:1]
+
+    assert f(shape) == jnp.ones(shape, dtype)[0, 1, 0:1]
+    assert f"tensor<{'x'.join(['?']*len(shape))}xcomplex<f64>>" in f.mlir
+
+
+def test_classical_tracing_2():
+    """Test that tensor primitive work in the classical tracing mode, the traced dimension case"""
+
+    @qjit
+    def f(x):
+        return jnp.ones(shape=[1, x], dtype=int)
+
+    assert_array_and_dtype_equal(f(3), jnp.ones((1, 3), dtype=int))
+
+
+@pytest.mark.skip("Dynamic arrays support in quantum control flow is not implemented")
+def test_quantum_tracing_1():
+    """Test that catalyst tensor primitive is compatible with quantum tracing mode"""
+
+    @qjit
+    @qml.qnode(qml.device("lightning.qubit", wires=4))
+    def f(shape):
+        i = 0
+        a = jnp.ones(shape, dtype=float)
+
+        @while_loop(lambda _, i: i < 3)
+        def loop(_, i):
+            qml.PauliX(wires=0)
+            b = jnp.ones(shape, dtype=float)
+            i += 1
+            return (b, i)
+
+        a2, _ = loop(a, i)
+        return a2
+
+    result = f([2, 3])
+    expected = jnp.ones([2, 3]) * 8
+    assert_array_and_dtype_equal(result, expected)
+
+
+@pytest.mark.skip("Dynamic arrays support in quantum control flow is not implemented")
+def test_quantum_tracing_2():
+    """Test that catalyst tensor primitive is compatible with quantum tracing mode"""
+
+    @qjit
+    @qml.qnode(qml.device("lightning.qubit", wires=4))
+    def f(x, y):
+        i = 0
+        a = jnp.ones((x, y + 1), dtype=float)
+
+        @while_loop(lambda _, i: i < 3)
+        def loop(_, i):
+            qml.PauliX(wires=0)
+            b = jnp.ones((x, y + 1), dtype=float)
+            i += 1
+            return (b, i)
+
+        a2, _ = loop(a, i)
+        return a2
+
+    result = f(2, 3)
+    print(result)
+    expected = jnp.ones((2, 4))
+    assert_array_and_dtype_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    "bad_shape",
+    [
+        [[2, 3]],
+        [2, 3.0],
+        [1, jnp.array(2, dtype=float)],
+    ],
+)
+def test_invalid_shapes(bad_shape):
+    """Test the unsupported shape formats"""
+
+    def f():
+        return jnp.empty(shape=bad_shape, dtype=int)
+
+    with pytest.raises(
+        TypeError, match="Shapes must be 1D sequences of concrete values of integer type"
+    ):
+        qjit(f)
+
+
+@pytest.mark.skip("Jax does not detect error in this use-case")
+def test_invalid_shapes_2():
+    """Test the unsupported shape formats"""
+    bad_shape = jnp.array([[3, 2]], dtype=int)
+
+    def f():
+        return jnp.empty(shape=bad_shape, dtype=int)
+
+    with pytest.raises(TypeError):
+        qjit(f)
+
+
+def test_shapes_type_conversion():
+    """Test fixes jax behavior regarding the shape conversions"""
+
+    def f(x):
+        return jnp.empty(shape=[2, x], dtype=int)
+
+    assert qjit(f)(3.1).shape == (2, 3)
+    assert qjit(f)(4.9).shape == (2, 4)
```

Comment:
This behaviour goes against JAX expected behaviour:

```
>>> jax.jit(lambda: jax.numpy.zeros([3.1]))()

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <lambda>
  File "/home/erick.ochoalopez/Code/env/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py", line 2171, in zeros
    shape = canonicalize_shape(shape)
  File "/home/erick.ochoalopez/Code/env/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py", line 80, in canonicalize_shape
    return core.canonicalize_shape(shape, context)  # type: ignore
TypeError: Shapes must be 1D sequences of concrete values of integer type, got [3.1].
```

I know we are not following all of JAX's decisions, but I am noting that a choice was made here to allow this behaviour vs doing a type check.

Reply:
Hmm. This test was intended to fix this strange behavior. I would leave it as-is for now, but it is good to know about this `canonicalize_shape` function. Are you ok if I mention it in this test's comment?

Reply:
(I have updated the description)

Reply:
Do we know why this passes for us though, if we are calling the same function (`jax.numpy.empty`)?
So at what point does jax invoke the check, and how are bypassing it?

Reply:
Added issue here: https://github.com/PennyLaneAI/catalyst/issues/378